### PR TITLE
Fix: Client memory leak

### LINKED ISSUE #300: LEAK: ByteBuf.release() was not called before it's garbage-collected.
zio-http version: 1.0.0.0-RC17
environment: Docker container built from adoptopenjdk/openjdk11:jre-11.0.11_9-alpine

This has showed up in my logs four times in the last few days. My code uses zio-http for both its client and server capabilities, and since the stack trace doesn't include any of my code, I don't know which of those two were the source of this.
```
{
   @timestamp: 2021-07-16T11:36:51.748+00:00
   level: ERROR
   logger_name: io.netty.util.ResourceLeakDetector
   message: LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:402)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	io.netty.channel.unix.PreferredDirectByteBufAllocator.ioBuffer(PreferredDirectByteBufAllocator.java:53)
	io.netty.channel.DefaultMaxMessagesRecvByteBufAllocator$MaxMessageHandle.allocate(DefaultMaxMessagesRecvByteBufAllocator.java:114)
	io.netty.channel.epoll.EpollRecvByteAllocatorHandle.allocate(EpollRecvByteAllocatorHandle.java:75)
	io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:780)
	io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
	io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Unknown Source)
   thread_name: epollEventLoopGroup-2-2
}
```

### PR DESCRIPTION
**Pull request by gciuloaica**

### CODE REVIEW

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 440617.96**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 438541.47**

File: zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala [hunk 1]
```diff
@@ -11,6 +12,8 @@ trait DecodeJResponse {
   def decodeJResponse(jRes: FullHttpResponse): Client.ClientResponse = {
     val status  = Status.fromHttpResponseStatus(jRes.status())
     val headers = Header.parse(jRes.headers())
-    Client.ClientResponse(status, headers, jRes.content())
+    val content =
+      Unpooled.buffer(jRes.content().readableBytes(), jRes.content().capacity()).writeBytes(jRes.content())
+    Client.ClientResponse(status, headers, content)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala:17]: As a general practice on the "backend" we don't use ZIO for writing any kind of server logic. The reason for doing that is primarily performance. ​What we can do in this case is move the logic to release content inside `decodeJResponse` instead.


The other things we might want to check is if the request doesn't have any content, don't waste CPU cycles in copying or releasing, instead use it as it is. Empty content AFAIK shouldn't cause leaks.

File: zio-http/src/main/scala/zhttp/service/client/ClientInboundHandler.scala [hunk 2]
```diff
@@ -18,7 +18,7 @@ final case class ClientInboundHandler[R](
 
   override def channelRead0(ctx: ChannelHandlerContext, msg: FullHttpResponse): Unit = {
     val clientResponse = decodeJResponse(msg)
-    zExec.unsafeRun(ctx)(promise.succeed(clientResponse))
+    zExec.unsafeRun(ctx)(promise.succeed(clientResponse).ensuring(Task(msg.release(msg.refCnt())).orDie))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/client/ClientInboundHandler.scala]: can drop the ensuring bit

File: zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala (hunk 1 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala:17]: - [x] Check if release is happening automatically
- [x] Optimize for perf by using the original content if it's empty.

Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala:17]: Release is not happening automatically.
@tusharmath please take a look again, I have updated the code for empty content case.

File: zio-http/src/main/scala/zhttp/service/client/ClientInboundHandler.scala (hunk 2 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/client/ClientInboundHandler.scala]: done

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 433388.40**

File: zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala [hunk 3]
```diff
@@ -11,6 +12,13 @@ trait DecodeJResponse {
   def decodeJResponse(jRes: FullHttpResponse): Client.ClientResponse = {
     val status  = Status.fromHttpResponseStatus(jRes.status())
     val headers = Header.parse(jRes.headers())
-    Client.ClientResponse(status, headers, jRes.content())
+    val content =
+      if (jRes.content().readableBytes() == 0) {
+        Unpooled.EMPTY_BUFFER
+      } else {
+        Unpooled.buffer(jRes.content().readableBytes(), jRes.content().capacity()).writeBytes(jRes.content())
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala]:
```suggestion
        Unpooled.copiedBuffer(jRes.content())
```
Will this have the same effect?

File: zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala [hunk 4]
```diff
@@ -11,6 +12,13 @@ trait DecodeJResponse {
...
+      }
+    jRes.release(jRes.refCnt())
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala:16]: Is this step required? I think writing the bytes should automatically release the direct memory.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 437555.54**

File: zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala (hunk 4 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala:16]: no, it is no releasing the memory

File: zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala (hunk 3 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/DecodeJResponse.scala]: yes, it will have effect and simplifies our code too

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 436036.01**

**Review by tusharmath — APPROVED** [review]

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 432643.52**

**Review by tusharmath — APPROVED** [review]

**Review by amitksingh1490 — APPROVED** [review]

**Pull request merged**